addons/test_mail/tests/test_mail_gateway.py

```diff
@@ -488,28 +488,89 @@ def test_message_process_alias_config_bounced_to(self):
         self.assertEqual(len(record), 1)
         res = record.get_metadata()[0].get('create_uid') or [None]
         self.assertEqual(res[0], self.user_employee.id)
         self.assertEqual(record.name, 'Specific2')
         self.assertFalse(record.custom_field)
 
+        self.assertEqual(self.alias.alias_status, 'valid')
+
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models', 'odoo.sql_db')
+    def test_message_process_alias_config_invalid_defaults(self):
+        """Sending a mail to a misconfigured alias must change its status to invalid + notify sender and responsible."""
+        test_model_track = self.env['ir.model']._get('mail.test.track')
+        container_custom_value_deleted = self.env['mail.test.container'].create({})
+        alias_valid = self.env['mail.alias'].create({
+            'alias_name': 'valid',
+            'alias_user_id': self.user_employee.id,
+            'alias_model_id': test_model_track.id,
+            'alias_contact': 'everyone',
+            'alias_defaults': f"{{'container_id': {container_custom_value_deleted.id}}}",
+        })
+
+        # Test that it works when the reference to container_id in alias default is not dangling.
+        self.assertEqual(alias_valid.alias_status, 'not_tested')
+        with self.mock_mail_gateway(), patch('odoo.addons.mail.models.mail_alias.Alias._set_alias_invalid',
+                                             autospec=True) as _set_alias_invalid_mock:
+            record = self.format_and_process(MAIL_TEMPLATE, self.email_from, 'valid@test.com', subject='Valid',
+                                             target_model=test_model_track.model)
+        _set_alias_invalid_mock.assert_not_called()
+        self.assertNotSentEmail()
+        self.assertEqual(record.container_id, container_custom_value_deleted)
+        self.assertEqual(alias_valid.alias_status, 'valid')
+
+        # Test with a dangling reference that must trigger bounce emails and set the alias status to invalid.
+        container_custom_value_deleted.unlink()
+        with self.assertRaises(Exception), patch('odoo.addons.mail.models.mail_alias.Alias._set_alias_invalid',
+                                                 autospec=True) as _set_alias_invalid_mock:
+            self.format_and_process(MAIL_TEMPLATE, self.email_from, 'valid@test.com', subject='Invalid',
+                                    target_model=test_model_track.model)
+
+        # method executed in another transaction, so we cannot test its result directly but just below
+        _set_alias_invalid_mock.assert_called_once()
+
+        # call notify_alias_invalid on the test transaction to validate its effect
+        alias, message, message_dict = _set_alias_invalid_mock.call_args.args
+        with self.mock_mail_gateway():
+            alias = self.env['mail.alias'].browse(alias.id)  # load alias in test transaction
+            alias._set_alias_invalid(message, message_dict)
```

Comment:
Do we have to call the method again ?

Reply:
Actually we assert that the mock has been called but the real method hasn’t yet. The idea is to call it to simulate a full execution and then test that the status of the alias has been changed.

The mock is necessary to avoid the real call during the execution of format_and_process because because it is called in another transaction not visible to the test. Calling it after in the same transaction allows to check everything.

Another solution would be to test the method _set_alias_invalid separately but it would make the test less comprehensible, I think. What do you think ?

Reply:
Ok then :)